File: apps/mobile/modules/joko-mobile-resource-pressure/android/src/main/java/app/joko/resourcepressure/JokoMobileResourcePressureModule.kt
package app.joko.resourcepressure

import android.content.ComponentCallbacks2
import android.content.Context
import android.content.res.Configuration
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class JokoMobileResourcePressureModule : Module(), ComponentCallbacks2 {
  private var applicationContext: Context? = null
  private var sequence = 0

  override fun definition() = ModuleDefinition {
    Name(MODULE_NAME)
    Events(RESOURCE_PRESSURE_EVENT)

    OnCreate {
      val context = appContext.reactContext?.applicationContext ?: return@OnCreate
      applicationContext = context
      context.registerComponentCallbacks(this@JokoMobileResourcePressureModule)
    }

    OnDestroy {
      applicationContext?.unregisterComponentCallbacks(this@JokoMobileResourcePressureModule)
      applicationContext = null
    }
  }

  override fun onConfigurationChanged(newConfig: Configuration) = Unit

  @Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
  override fun onLowMemory() {
    emitPressure("critical")
  }

  @Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
  override fun onTrimMemory(level: Int) {
    if (level == ComponentCallbacks2.TRIM_MEMORY_UI_HIDDEN
      || level < ComponentCallbacks2.TRIM_MEMORY_RUNNING_MODERATE) return
    val severity = if (level >= ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL) "critical" else "warning"
    emitPressure(severity)
  }

  private fun emitPressure(severity: String) {
    sequence = if (sequence == Int.MAX_VALUE) 1 else sequence + 1
    sendEvent(RESOURCE_PRESSURE_EVENT, mapOf(
      "sequence" to sequence,
      "severity" to severity,
      "platform" to "android"
    ))
  }

  private companion object {
    const val MODULE_NAME = "JokoMobileResourcePressure"
    const val RESOURCE_PRESSURE_EVENT = "onResourcePressure"
  }
}

File: apps/mobile/modules/joko-mobile-resource-pressure/ios/JokoMobileResourcePressureModule.swift
import ExpoModulesCore
import UIKit

public final class JokoMobileResourcePressureModule: Module {
  private var observer: NSObjectProtocol?
  private var sequence = 0

  public func definition() -> ModuleDefinition {
    Name("JokoMobileResourcePressure")
    Events("onResourcePressure")

    OnCreate {
      self.observer = NotificationCenter.default.addObserver(
        forName: UIApplication.didReceiveMemoryWarningNotification,
        object: nil,
        queue: .main
      ) { [weak self] _ in
        self?.emitPressure()
      }
    }

    OnDestroy {
      if let observer = self.observer {
        NotificationCenter.default.removeObserver(observer)
      }
      self.observer = nil
    }
  }

  private func emitPressure() {
    sequence = sequence == Int.max ? 1 : sequence + 1
    sendEvent("onResourcePressure", [
      "sequence": sequence,
      "severity": "critical",
      "platform": "ios"
    ])
  }
}
